src/main/java/com/mathworks/ci/TestResultsViewAction.java

```diff
@@ -0,0 +1,265 @@
+        List<List<TestFile>> testResults = new ArrayList<>();
+        FilePath fl = new FilePath(new File(build.getRootDir().getAbsolutePath() + File.separator + MatlabBuilderConstants.TEST_RESULTS_VIEW_ARTIFACT + this.actionID + ".json"));
+        try (InputStreamReader reader = new InputStreamReader(Files.newInputStream(Paths.get(fl.toURI())), StandardCharsets.UTF_8)) {
+            totalCount = 0;
+            passedCount = 0;
+            failedCount = 0;
+            incompleteCount = 0;
+            notRunCount = totalCount;
+
+            JSONArray testArtifact = (JSONArray) new JSONParser().parse(reader);
+            Iterator<JSONArray> testArtifactIterator = testArtifact.iterator();
+
+            while(testArtifactIterator.hasNext()){
+                Object jsonTestSessionResults = testArtifactIterator.next();
+
+                List<TestFile> testSessionResults = new ArrayList<>();
+                Map<String, TestFile> map = new HashMap<>();
+
+                if(jsonTestSessionResults instanceof JSONArray){
+                    JSONArray jsonTestSessionResultsArray = (JSONArray) jsonTestSessionResults;
+                    Iterator<JSONObject> testSessionResultsIterator = jsonTestSessionResultsArray.iterator();
+
+                    while(testSessionResultsIterator.hasNext()){
+                        JSONObject jsonTestCase = testSessionResultsIterator.next();
+                        getTestSessionResults(testSessionResults, jsonTestCase, map);
+                    }
+                }
+                else if(jsonTestSessionResults instanceof JSONObject) {
+                    JSONObject jsonTestCase = (JSONObject) jsonTestSessionResults;
+                    getTestSessionResults(testSessionResults, jsonTestCase, map);
+                }
+
+                testResults.add(testSessionResults);
+            }
+        }
+        catch (Exception e) {
+            throw new IOException(e.getLocalizedMessage());
+        }
+
+        return testResults;
+    }
+
+    private void getTestSessionResults(List<TestFile> testSessionResults, JSONObject jsonTestCase, Map<String, TestFile> map) throws IOException, InterruptedException {
+        FilePath baseFolder = new FilePath(new File(jsonTestCase.get("BaseFolder").toString()));
+        JSONObject testCaseResult = (JSONObject) jsonTestCase.get("TestResult");
+
+        // Not OS dependent
+        String[] testNameSplit = testCaseResult.get("Name").toString().split("/");
+        String testFileName = testNameSplit[0];
+        String testCaseName = testNameSplit[1];
+
+        TestFile testFile = map.get(baseFolder + File.separator + testFileName);
+        if(testFile == null) {
+            testFile = new TestFile();
+            testFile.setName(testFileName);
+
+            map.put(baseFolder + File.separator + testFileName, testFile);
+            testSessionResults.add(testFile);

```

Comment:
If a test file is in multiple sessions, it will get added to all respective `testSessionResults` arrays. At line 80, you can see that the `testSessionResults` array gets initialised at the beginning of when we start reading results of a particular test session.
This `if` block makes sure within a `testSessionResults` array, we have exactly one instance of each test file that ran in the particular test session. Multiple test cases of the same test file are stored within the `MatlabTestFile` object as a list of `MatlabTestCase`.